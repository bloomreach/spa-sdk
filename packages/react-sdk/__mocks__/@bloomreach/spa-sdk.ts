/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const spaSdk = jest.requireActual('@bloomreach/spa-sdk');

export const META_POSITION_BEGIN = spaSdk.META_POSITION_BEGIN;
export const META_POSITION_END = spaSdk.META_POSITION_END;

export function mockMeta(data: string, position: string) {
  return {
    getData: jest.fn().mockReturnValue(data),
    getPosition: jest.fn().mockReturnValue(position),
  };
}

export function mockNoCommentMeta() {
  return mockMeta('not-a-comment', META_POSITION_BEGIN);
}

const componentMock = {
  getMeta: jest.fn().mockReturnValue([
    mockMeta('meta-begin', META_POSITION_BEGIN),
    mockMeta('meta-end', META_POSITION_END),
  ]),
};

export const pageMock = {
  getComponent: jest.fn().mockReturnValue(componentMock),
  getContent: jest.fn(),
  getTitle: jest.fn(),
  sync: jest.fn(),
};

export const initialize = jest.fn();
export const destroy = jest.fn();

export function isMetaComment(value: any) {
  return value.getData() !== 'not-a-comment';
}
