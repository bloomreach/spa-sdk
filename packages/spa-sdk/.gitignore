/coverage
/docs
/lib
/node_modules
.DS_store