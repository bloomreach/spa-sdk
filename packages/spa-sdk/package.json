{
  "name": "@bloomreach/spa-sdk",
  "version": "14.0.0",
  "description": "Bloomreach SPA SDK",
  "keywords": [
    "bloomreach",
    "spa",
    "sdk"
  ],
  "author": "Bloomreach B.V.",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "https://code.onehippo.org/cms-community/bloomreach-spa-sdk",
    "directory": "packages/spa-sdk"
  },
  "bugs": {
    "url": "https://issues.onehippo.com/projects/CMS/issues"
  },
  "main": "dist/index.js",
  "browser": "dist/index.js",
  "module": "dist/index.mjs",
  "es2015": "dist/index.es6.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist/"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js",
    "docs": "tsc --emitDeclarationOnly && dts-bundle-generator --out-file dist/spa-sdk.d.ts dist/index.d.ts && typedoc dist/spa-sdk.d.ts && minicat src/docs/typedoc.css >> docs/assets/css/main.css && rm dist/spa-sdk.d.ts",
    "lint": "tslint --project .",
    "test": "jest"
  },
  "jest": {
    "testEnvironment": "jsdom",
    "coverageDirectory": "./coverage",
    "collectCoverage": true,
    "collectCoverageFrom": [
      "src/**/*.ts"
    ],
    "moduleFileExtensions": [
      "ts",
      "js"
    ],
    "transform": {
      "^.+\\.ts$": "ts-jest"
    },
    "globals": {
      "ts-jest": {
        "tsConfig": "tsconfig.json"
      }
    },
    "testMatch": [
      "**/?(*.)spec.ts"
    ]
  },
  "devDependencies": {
    "@babel/core": "^7.6",
    "@babel/plugin-proposal-class-properties": "^7.5",
    "@babel/plugin-proposal-object-rest-spread": "^7.5",
    "@babel/plugin-transform-object-assign": "^7.2",
    "@babel/polyfill": "^7.6",
    "@babel/preset-env": "^7.6",
    "@types/jest": "^24.0",
    "@types/node": "^12.0",
    "babel-loader": "^8.0",
    "babel-plugin-transform-async-to-promises": "^0.8",
    "dts-bundle-generator": "^3.2",
    "jest": "^24.0",
    "jest-environment-jsdom-fifteen": "^1.0",
    "js-beautify": "^1.0",
    "minicat": "^1.0",
    "rollup": "^1.21",
    "rollup-plugin-babel": "^4.3",
    "rollup-plugin-terser": "^5.1",
    "rollup-plugin-typescript2": "^0.24",
    "ts-jest": "^24.0",
    "ts-loader": "^5.0",
    "tslint": "^5.0",
    "tslint-config-airbnb": "^5.0",
    "tslint-loader": "^3.5",
    "typedoc": "^0.15.0",
    "typescript": "^3.6"
  },
  "dependencies": {
    "emittery": "^0.4"
  }
}
