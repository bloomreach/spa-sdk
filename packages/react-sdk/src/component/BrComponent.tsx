/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { isComponent, Component } from '@bloomreach/spa-sdk';
import { BrComponentContext } from './BrComponentContext';
import { BrNode } from './BrNode';

interface BrComponentProps {
  path?: string;
}

interface BrComponentState {
  components: Component[];
}

export class BrComponent extends React.Component<BrComponentProps, BrComponentState> {
  static contextType = BrComponentContext;

  constructor(props: BrComponentProps, public context: React.ContextType<typeof BrComponentContext>) {
    super(props, context);

    this.state = {
      components: this.getComponents(),
    };
  }

  componentDidUpdate(prevProps: BrComponentProps) {
    if (this.props.path !== prevProps.path) {
      this.setState({ components: this.getComponents() });
    }
  }

  private getComponents() {
    if (!this.context) {
      return [];
    }
    if (!this.props.path) {
      return this.context.getChildren();
    }

    const component = this.context.getComponent(...this.props.path.split('/'));

    return component ? [component] : [];
  }

  private renderComponents() {
    return this.state.components.map((component, index) => (
      <BrNode key={index} component={component}>{this.props.children}</BrNode>
    ));
  }

  render() {
    return <>{this.renderComponents()}</>;
  }
}
