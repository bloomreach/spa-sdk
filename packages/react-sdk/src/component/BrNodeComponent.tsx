/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component } from '@bloomreach/spa-sdk';
import React, { useContext } from 'react';
import { BrMeta } from '../meta';
import { BrMappingContext } from './BrMappingContext';
import { BrProps } from './BrProps';

export function BrNodeComponent<T extends Component>(props: React.PropsWithChildren<BrProps<T>>): React.ReactElement {
  const context = useContext(BrMappingContext);
  const { component, children } = props;

  const mapping = component && (context[component.getName()] as React.ComponentType<BrProps>);
  const meta = component?.getMeta();

  const content = mapping ? React.createElement(mapping, props) : children;

  return React.createElement(BrMeta, { meta }, content);
}
