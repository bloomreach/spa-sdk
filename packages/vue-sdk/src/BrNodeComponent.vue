<!--
  Copyright 2020 Hippo B.V. (http://www.onehippo.com)

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
  -->

<template>
  <br-meta :meta="component.getMeta()">
    <slot>
      <br-node-container-item v-if="isContainerItem(component)" />

      <div v-else-if="isContainer(component)">
        <br-node-component v-for="(component, key) in component.getChildren()" :key="key" :component="component" />
      </div>

      <br-node-component v-else v-for="(component, key) in component.getChildren()" :key="key" :component="component" />
    </slot>
  </br-meta>
</template>

<script lang="ts">
import { Component as SpaComponent, Page, isContainerItem, isContainer } from '@bloomreach/spa-sdk';
import { Component, InjectReactive, Prop, Provide, Vue } from 'vue-property-decorator';
import BrMeta from './BrMeta.vue';
import BrNodeContainerItem from './BrNodeContainerItem.vue';

@Component({
  components: {
    BrMeta,
    BrNodeContainerItem,
  },
  methods: {
    isContainerItem,
    isContainer,
  },
})
export default class BrNodeComponent extends Vue {
  @Prop() component!: SpaComponent;

  @InjectReactive() private page!: Page;

  @Provide() // ProvideReactive doesn't work with recursive components
  private component$() {
    return this.component;
  }
}
</script>
