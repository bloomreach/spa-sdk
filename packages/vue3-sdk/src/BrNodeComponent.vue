<!--
  - Copyright 2023 Bloomreach
  -
  - Licensed under the Apache License, Version 2.0 (the "License");
  - you may not use this file except in compliance with the License.
  - You may obtain a copy of the License at
  -
  -   http://www.apache.org/licenses/LICENSE-2.0
  -
  - Unless required by applicable law or agreed to in writing, software
  - distributed under the License is distributed on an "AS IS" BASIS,
  - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  - See the License for the specific language governing permissions and
  - limitations under the License.
  -->

<template>
  <br-meta :meta="meta">
    <slot>
      <br-node-container-item v-if="isContainerItem(component)"/>

      <br-node-container v-else-if="isContainer(component)">
        <br-node-component v-for="(component, key) in children" :key="key" :component="component"/>
      </br-node-container>

      <component v-else-if="name && name in mapping" :is="mapping[name]" :component="component" :page="page"/>

      <br-node-component v-else v-for="(component, key) in children" :key="key" :component="component"/>
    </slot>
  </br-meta>
</template>

<script setup lang="ts">
import BrMeta from '@/BrMeta.vue';
import BrNodeContainer from '@/BrNodeContainer.vue';
import BrNodeContainerItem from '@/BrNodeContainerItem.vue';
import { component$, mapping$, page$ } from '@/providerKeys';
import type { Component, Page } from '@bloomreach/spa-sdk';
import { isContainer, isContainerItem } from '@bloomreach/spa-sdk';
import { inject, provide } from 'vue';
import type { Ref } from 'vue';

const { component } = defineProps<{ component?: Component }>();
const page = inject(page$, {} as Ref<Page>);
const mapping = inject(mapping$, {});
const children = component?.getChildren();
const meta = component?.getMeta();
const name = component?.getName();

provide(component$, component);
</script>
