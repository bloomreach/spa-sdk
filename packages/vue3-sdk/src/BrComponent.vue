<!--
  Copyright 2023 Bloomreach

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

    https://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
  -->

<template>
  <br-node-component v-for="component in components" :component="component">
    <slot :component="component" :page="page"></slot>
  </br-node-component>
</template>

<script setup lang="ts">
import BrNodeComponent from '@/BrNodeComponent.vue';
import { component$, page$ } from '@/providerKeys';
import type { Component } from '@bloomreach/spa-sdk';
import { isComponent } from '@bloomreach/spa-sdk';
import { computed, inject } from 'vue';

function getComponents(
  parentComponent: Component | undefined,
  propsComponent: Component | string | undefined,
): Component[] {
  if (isComponent(propsComponent)) {
    return [propsComponent];
  }

  if (!parentComponent) {
    return [];
  }

  if (!propsComponent) {
    return parentComponent.getChildren();
  }

  const parentComponents = parentComponent.getComponent(...propsComponent.split('/'));
  return parentComponents ? [parentComponents] : [];
}

const props = defineProps<{ component?: Component | string }>();
const page = inject(page$);
const parent = inject(component$);
const components = computed(() => getComponents(parent?.value, props.component));
</script>
