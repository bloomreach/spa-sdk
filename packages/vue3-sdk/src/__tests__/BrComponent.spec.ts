/*
 * Copyright 2020-2023 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import BrComponent from '@/BrComponent.vue';
import BrNodeComponent from '@/BrNodeComponent.vue';
import type { Component, Page } from '@bloomreach/spa-sdk';
import { isComponent } from '@bloomreach/spa-sdk';
import { shallowMount } from '@vue/test-utils';
import type { GlobalMountOptions } from '@vue/test-utils/dist/types';
import type { Mocked } from 'vitest';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { ref } from 'vue';
import { component$, page$ } from '../providerKeys';

vi.mock('@bloomreach/spa-sdk');

describe('BrComponent', () => {
  let parent: Mocked<Component>;
  let page: Mocked<Page>;
  let global: GlobalMountOptions;

  beforeEach(() => {
    parent = {
      getChildren: vi.fn(() => []),
      getComponent: vi.fn(),
    } as unknown as typeof parent;
    page = { sync: vi.fn() } as unknown as typeof page;
    global = {
      provide: {
        [component$ as symbol]: ref(parent),
        [page$ as symbol]: ref(page),
      },
    };
  });

  describe('render', () => {
    it('should render a single component', () => {
      vi.mocked(isComponent).mockReturnValueOnce(true);
      const component = 'some component';
      const wrapper = shallowMount(BrComponent, {
        global,
        props: { component },
      });
      const props = wrapper.findComponent(BrNodeComponent).props();

      expect(wrapper.html()).toMatchSnapshot();
      expect(props.component).toStrictEqual(component);
    });

    it('should render children if the component prop is omitted', () => {
      const child1 = {} as Component;
      const child2 = {} as Component;
      parent.getChildren.mockReturnValueOnce([child1, child2]);
      const wrapper = shallowMount(BrComponent, {
        global,
      });
      const children = wrapper.findAllComponents(BrNodeComponent);

      expect(children.at(0)?.props().component).toBe(child1);
      expect(children.at(1)?.props().component).toBe(child2);
    });

    it('should render a component by path', () => {
      const component = {} as Component;
      parent.getComponent.mockReturnValueOnce(component);
      const wrapper = shallowMount(BrComponent, {
        global,
        props: { component: 'a/b' },
      });
      const props = wrapper.findComponent(BrNodeComponent).props();

      expect(parent.getComponent).toBeCalledWith('a', 'b');
      expect(props.component).toBe(component);
    });

    it('should render nothing if no component found', () => {
      const wrapper = shallowMount(BrComponent, {
        global,
        props: { component: 'a/b' },
      });

      expect(wrapper.html()).toBe('');
    });
  });
});
