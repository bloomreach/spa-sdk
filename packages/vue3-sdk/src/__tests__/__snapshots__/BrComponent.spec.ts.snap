// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BrComponent > render > should render a single component 1`] = `"<br-node-component-stub component=\\"some component\\"></br-node-component-stub>"`;
